File: src/main/java/com/enterprise/iqk/service/TenantCostService.java
package com.enterprise.iqk.service;

import com.enterprise.iqk.config.properties.CostGovernanceProperties;
import com.enterprise.iqk.domain.TenantBudget;
import com.enterprise.iqk.domain.vo.TenantBudgetUpdateVO;
import com.enterprise.iqk.domain.vo.TenantCostSummaryVO;
import com.enterprise.iqk.mapper.TenantBudgetMapper;
import com.enterprise.iqk.mapper.TenantUsageDailyMapper;
import com.enterprise.iqk.security.TenantContext;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.util.StringUtils;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.YearMonth;
import java.util.Locale;

@Service
@RequiredArgsConstructor
public class TenantCostService {
    private final TenantBudgetMapper tenantBudgetMapper;
    private final TenantUsageDailyMapper tenantUsageDailyMapper;
    private final CostGovernanceProperties costGovernanceProperties;

    public void assertBudget(String tenantId, String costTier, long inputTokens, long outputTokens) {
        if (!costGovernanceProperties.isEnabled()) {
            return;
        }
        String tenant = TenantContext.normalize(tenantId);
        TenantBudget budget = ensureBudget(tenant);
        BigDecimal estimatedCost = calculateCost(costTier, inputTokens + outputTokens);
        BigDecimal monthCost = monthCost(tenant, YearMonth.now());
        BigDecimal projected = monthCost.add(estimatedCost);
        boolean hardLimit = budget.getHardLimitEnabled() != null && budget.getHardLimitEnabled() == 1;
        if (hardLimit && projected.compareTo(defaultDecimal(budget.getMonthlyBudgetUsd())) > 0) {
            throw new IllegalArgumentException("tenant budget exceeded, request blocked");
        }
    }

    public void recordUsage(String tenantId,
                            String costTier,
                            long inputTokens,
                            long outputTokens,
                            String endpointTag) {
        if (!costGovernanceProperties.isEnabled()) {
            return;
        }
        String tenant = TenantContext.normalize(tenantId);
        long safeInput = Math.max(0, inputTokens);
        long safeOutput = Math.max(0, outputTokens);
        BigDecimal cost = calculateCost(costTier, safeInput + safeOutput);
        tenantUsageDailyMapper.addUsage(
                tenant,
                LocalDate.now(),
                1,
                safeInput,
                safeOutput,
                cost
        );
    }

    public TenantCostSummaryVO summary(String tenantId) {
        String tenant = TenantContext.normalize(tenantId);
        TenantBudget budget = ensureBudget(tenant);
        YearMonth now = YearMonth.now();
        LocalDate monthStart = now.atDay(1);
        LocalDate monthEnd = now.atEndOfMonth();
        LocalDate today = LocalDate.now();

        Long monthRequestCount = safeLong(tenantUsageDailyMapper.sumRequestCount(tenant, monthStart, monthEnd));
        Long monthInput = safeLong(tenantUsageDailyMapper.sumInputTokens(tenant, monthStart, monthEnd));
        Long monthOutput = safeLong(tenantUsageDailyMapper.sumOutputTokens(tenant, monthStart, monthEnd));
        BigDecimal monthCost = defaultDecimal(tenantUsageDailyMapper.sumCostUsd(tenant, monthStart, monthEnd));

        Long todayRequestCount = safeLong(tenantUsageDailyMapper.sumRequestCount(tenant, today, today));
        BigDecimal todayCost = defaultDecimal(tenantUsageDailyMapper.sumCostUsd(tenant, today, today));

        BigDecimal budgetLimit = defaultDecimal(budget.getMonthlyBudgetUsd());
        BigDecimal remaining = budgetLimit.subtract(monthCost).max(BigDecimal.ZERO);
        boolean exceeded = monthCost.compareTo(budgetLimit) > 0;

        return TenantCostSummaryVO.builder()
                .tenantId(tenant)
                .month(now.toString())
                .monthlyBudgetUsd(budgetLimit)
                .hardLimitEnabled(budget.getHardLimitEnabled() != null && budget.getHardLimitEnabled() == 1)
                .monthCostUsd(scale(monthCost))
                .monthRequestCount(monthRequestCount)
                .monthInputTokens(monthInput)
                .monthOutputTokens(monthOutput)
                .todayCostUsd(scale(todayCost))
                .todayRequestCount(todayRequestCount)
                .budgetRemainingUsd(scale(remaining))
                .budgetExceeded(exceeded)
                .build();
    }

    public TenantCostSummaryVO updateBudget(TenantBudgetUpdateVO request) {
        if (request == null) {
            throw new IllegalArgumentException("budget payload is required");
        }
        String tenant = TenantContext.normalize(request.getTenantId());
        TenantBudget budget = ensureBudget(tenant);
        if (request.getMonthlyBudgetUsd() != null) {
            if (request.getMonthlyBudgetUsd().compareTo(BigDecimal.ZERO) < 0) {
                throw new IllegalArgumentException("monthlyBudgetUsd must be non-negative");
            }
            budget.setMonthlyBudgetUsd(request.getMonthlyBudgetUsd());
        }
        if (request.getHardLimitEnabled() != null) {
            budget.setHardLimitEnabled(Boolean.TRUE.equals(request.getHardLimitEnabled()) ? 1 : 0);
        }
        budget.setUpdatedAt(LocalDateTime.now());
        tenantBudgetMapper.updateById(budget);
        return summary(tenant);
    }

    public long estimateTokens(String text) {
        if (!StringUtils.hasText(text)) {
            return 0;
        }
        int divisor = Math.max(1, costGovernanceProperties.getTokenEstimateDivisor());
        int length = text.codePointCount(0, text.length());
        return Math.max(1L, (length + divisor - 1L) / divisor);
    }

    private TenantBudget ensureBudget(String tenantId) {
        TenantBudget existing = tenantBudgetMapper.findByTenantId(tenantId);
        if (existing != null) {
            return existing;
        }
        LocalDateTime now = LocalDateTime.now();
        TenantBudget inserted = TenantBudget.builder()
                .tenantId(tenantId)
                .monthlyBudgetUsd(defaultDecimal(costGovernanceProperties.getDefaultMonthlyBudgetUsd()))
                .hardLimitEnabled(costGovernanceProperties.isDefaultHardLimitEnabled() ? 1 : 0)
                .createdAt(now)
                .updatedAt(now)
                .build();
        tenantBudgetMapper.insert(inserted);
        return tenantBudgetMapper.findByTenantId(tenantId);
    }

    private BigDecimal monthCost(String tenantId, YearMonth month) {
        BigDecimal total = tenantUsageDailyMapper.sumCostUsd(tenantId, month.atDay(1), month.atEndOfMonth());
        return defaultDecimal(total);
    }

    private BigDecimal calculateCost(String costTier, long totalTokens) {
        String tier = StringUtils.hasText(costTier) ? costTier.trim().toLowerCase(Locale.ROOT) : "medium";
        BigDecimal unit = costGovernanceProperties.getUsdPer1kTokens().getOrDefault(
                tier,
                costGovernanceProperties.getUsdPer1kTokens().getOrDefault("medium", new BigDecimal("0.0030"))
        );
        BigDecimal tokens = BigDecimal.valueOf(Math.max(0, totalTokens));
        return tokens.multiply(defaultDecimal(unit))
                .divide(BigDecimal.valueOf(1000), 6, RoundingMode.HALF_UP);
    }

    private Long safeLong(Long value) {
        return value == null ? 0L : value;
    }

    private BigDecimal defaultDecimal(BigDecimal value) {
        return value == null ? BigDecimal.ZERO : value;
    }

    private BigDecimal scale(BigDecimal value) {
        return defaultDecimal(value).setScale(4, RoundingMode.HALF_UP);
    }
}

File: typescript/apps/api/src/platform/tenant-cost.service.ts
import { Injectable } from "@nestjs/common";

import { nowIso } from "../common/ids.js";
import { normalizeTenant } from "../common/tenant.js";
import { env } from "../config/env.js";
import { PlatformStore, tenantUsageKey } from "./platform.store.js";

export interface TenantCostSummary {
  tenantId: string;
  month: string;
  monthlyBudgetUsd: number;
  hardLimitEnabled: boolean;
  monthCostUsd: number;
  monthRequestCount: number;
  monthInputTokens: number;
  monthOutputTokens: number;
  todayCostUsd: number;
  todayRequestCount: number;
  budgetRemainingUsd: number;
  budgetExceeded: boolean;
}

@Injectable()
export class TenantCostService {
  constructor(private readonly store: PlatformStore) {}

  assertBudget(tenantId: string | undefined, costTier: string | undefined, inputTokens: number, outputTokens: number): void {
    if (!env.APP_COST_ENABLED) {
      return;
    }
    const tenant = normalizeTenant(tenantId);
    const budget = this.ensureBudget(tenant);
    const projected = this.monthCost(tenant, currentMonth()) + this.calculateCost(costTier, inputTokens + outputTokens);
    if (budget.hardLimitEnabled && projected > budget.monthlyBudgetUsd) {
      throw new Error("tenant budget exceeded, request blocked");
    }
  }

  recordUsage(tenantId: string | undefined, costTier: string | undefined, inputTokens: number, outputTokens: number): void {
    if (!env.APP_COST_ENABLED) {
      return;
    }
    const tenant = normalizeTenant(tenantId);
    this.ensureBudget(tenant);
    const today = new Date().toISOString().slice(0, 10);
    const key = tenantUsageKey(tenant, today);
    const existing = this.store.tenantUsageDaily.get(key);
    const now = nowIso();
    const usage = existing ?? {
      tenantId: tenant,
      usageDate: today,
      requestCount: 0,
      inputTokens: 0,
      outputTokens: 0,
      totalCostUsd: 0,
      createdAt: now,
      updatedAt: now
    };
    usage.requestCount += 1;
    usage.inputTokens += Math.max(0, inputTokens);
    usage.outputTokens += Math.max(0, outputTokens);
    usage.totalCostUsd = roundMoney(usage.totalCostUsd + this.calculateCost(costTier, inputTokens + outputTokens), 6);
    usage.updatedAt = now;
    this.store.tenantUsageDaily.set(key, usage);
    this.store.persist();
  }

  summary(tenantId: string | undefined): TenantCostSummary {
    const tenant = normalizeTenant(tenantId);
    const budget = this.ensureBudget(tenant);
    const month = currentMonth();
    const today = new Date().toISOString().slice(0, 10);
    const monthRows = [...this.store.tenantUsageDaily.values()]
      .filter((usage) => usage.tenantId === tenant && usage.usageDate.startsWith(month));
    const todayRows = monthRows.filter((usage) => usage.usageDate === today);
    const monthCostUsd = roundMoney(sum(monthRows.map((usage) => usage.totalCostUsd)), 4);
    const todayCostUsd = roundMoney(sum(todayRows.map((usage) => usage.totalCostUsd)), 4);
    const budgetRemainingUsd = roundMoney(Math.max(0, budget.monthlyBudgetUsd - monthCostUsd), 4);
    return {
      tenantId: tenant,
      month,
      monthlyBudgetUsd: budget.monthlyBudgetUsd,
      hardLimitEnabled: budget.hardLimitEnabled,
      monthCostUsd,
      monthRequestCount: sum(monthRows.map((usage) => usage.requestCount)),
      monthInputTokens: sum(monthRows.map((usage) => usage.inputTokens)),
      monthOutputTokens: sum(monthRows.map((usage) => usage.outputTokens)),
      todayCostUsd,
      todayRequestCount: sum(todayRows.map((usage) => usage.requestCount)),
      budgetRemainingUsd,
      budgetExceeded: monthCostUsd > budget.monthlyBudgetUsd
    };
  }

  updateBudget(request: { tenantId?: string; monthlyBudgetUsd?: number; hardLimitEnabled?: boolean }): TenantCostSummary {
    const tenant = normalizeTenant(request.tenantId);
    const existing = this.ensureBudget(tenant);
    if (request.monthlyBudgetUsd !== undefined && request.monthlyBudgetUsd < 0) {
      throw new Error("monthlyBudgetUsd must be non-negative");
    }
    this.store.tenantBudgets.set(tenant, {
      tenantId: tenant,
      monthlyBudgetUsd: request.monthlyBudgetUsd ?? existing.monthlyBudgetUsd,
      hardLimitEnabled: request.hardLimitEnabled ?? existing.hardLimitEnabled,
      createdAt: existing.createdAt,
      updatedAt: nowIso()
    });
    this.store.persist();
    return this.summary(tenant);
  }

  estimateTokens(text: string | undefined): number {
    if (!text?.trim()) {
      return 0;
    }
    const length = [...text].length;
    return Math.max(1, Math.ceil(length / env.APP_COST_TOKEN_ESTIMATE_DIVISOR));
  }

  calculateCost(costTier: string | undefined, totalTokens: number): number {
    const tier = (costTier ?? "balanced").toLowerCase();
    const unit = tier === "high" || tier === "quality"
      ? env.APP_COST_USD_PER_1K_HIGH
      : tier === "low" || tier === "economy"
        ? env.APP_COST_USD_PER_1K_LOW
        : env.APP_COST_USD_PER_1K_BALANCED;
    return roundMoney(Math.max(0, totalTokens) * unit / 1000, 6);
  }

  private ensureBudget(tenantId: string) {
    const existing = this.store.tenantBudgets.get(tenantId);
    if (existing) {
      return existing;
    }
    const now = nowIso();
    const budget = {
      tenantId,
      monthlyBudgetUsd: env.APP_COST_DEFAULT_MONTHLY_BUDGET_USD,
      hardLimitEnabled: env.APP_COST_DEFAULT_HARD_LIMIT_ENABLED,
      createdAt: now,
      updatedAt: now
    };
    this.store.tenantBudgets.set(tenantId, budget);
    this.store.persist();
    return budget;
  }

  private monthCost(tenantId: string, month: string): number {
    return sum([...this.store.tenantUsageDaily.values()]
      .filter((usage) => usage.tenantId === tenantId && usage.usageDate.startsWith(month))
      .map((usage) => usage.totalCostUsd));
  }
}

function currentMonth(): string {
  return new Date().toISOString().slice(0, 7);
}

function sum(values: number[]): number {
  return values.reduce((total, value) => total + value, 0);
}

function roundMoney(value: number, digits: number): number {
  const factor = 10 ** digits;
  return Math.round(value * factor) / factor;
}
